overlord/configstate/configcore/homedirs_test.go

```diff
@@ -0,0 +1,216 @@
+ *
+ */
+
+package configcore_test
+
+import (
+	"errors"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	. "gopkg.in/check.v1"
+
+	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/overlord/configstate/configcore"
+)
+
+type mockedFileInfo struct {
+	isDir bool
+}
+
+func (f *mockedFileInfo) Name() string {
+	return ""
+}
+
+func (f *mockedFileInfo) Size() int64 {
+	return 0
+}
+
+func (f *mockedFileInfo) Mode() os.FileMode {
+	return 0
+}
+
+func (f *mockedFileInfo) ModTime() time.Time {
+	return time.Unix(0, 0)
+}
+
+func (f *mockedFileInfo) IsDir() bool {
+	return f.isDir
+}
+
+func (f *mockedFileInfo) Sys() interface{} {
+	return nil
+}
+
+type homedirsSuite struct {
+	configcoreSuite
+}
+
+var _ = Suite(&homedirsSuite{})
+
+func (s *homedirsSuite) SetUpTest(c *C) {
+	s.configcoreSuite.SetUpTest(c)
+
+	etcDir := filepath.Join(dirs.GlobalRootDir, "/etc/")
+	err := os.MkdirAll(etcDir, 0755)
+	c.Assert(err, IsNil)
+	s.AddCleanup(func() {
+		err := os.RemoveAll(etcDir)
+		c.Assert(err, IsNil)
+	})
+
+	restore := configcore.MockStat(func(path string) (os.FileInfo, error) {
+		if strings.HasPrefix(path, "/home/existingDir") {
+			return &mockedFileInfo{isDir: true}, nil
+		} else if strings.HasPrefix(path, "/home/existingFile") {
+			return &mockedFileInfo{isDir: false}, nil
+		} else {
+			return nil, errors.New("stat failed")
+		}
+	})
+	s.AddCleanup(restore)
+
+}
+
+func (s *homedirsSuite) TestValidationUnhappy(c *C) {
+	for _, testData := range []struct {
+		homedirs      string
+		expectedError string
+	}{
+		{"./here", `path "\./here" is not absolute`},
+		// empty path in list
+		{",/home", `path "" is not absolute`},
+		{"/somewhere/else", `path "/somewhere/else/" unsupported: must start with one of: /home/`},
+		{"/home/foo[12]", `home path invalid: "/home/foo\[12\]" contains a reserved apparmor char.*`},
+		{"/lib/homes", `path "/lib/homes/" uses reserved root directory "/lib/"`},
+		{"/home/missing", `cannot get directory info for "/home/missing/".*`},
+		{"/home/existingFile", `path "/home/existingFile/" is not a directory`},
+	} {
+		err := configcore.Run(coreDev, &mockConf{
+			state: s.state,
+			conf: map[string]interface{}{
+				"homedirs": testData.homedirs,
+			},
+		})
+		c.Assert(err, ErrorMatches, testData.expectedError, Commentf("%v", testData.homedirs))
+	}
+}
+
+func (s *homedirsSuite) TestConfigureOpenFailure(c *C) {
+	var systemInfoPath string
+	restore := configcore.MockOpenFile(func(path string, flags int, mode os.FileMode) (*os.File, error) {
+		systemInfoPath = path
+		return nil, errors.New("open failure")
+	})
+	defer restore()
+	err := configcore.Run(coreDev, &mockConf{
+		state: s.state,
+		conf: map[string]interface{}{
+			"homedirs": "/home/existingDir",
+		},
+	})
+	expectedPath := filepath.Join(dirs.SnapdStateDir(dirs.GlobalRootDir), "system-params")
+	c.Check(systemInfoPath, Equals, expectedPath)
+	c.Assert(err, ErrorMatches, "open failure")
```

Comment:
(nitpick^9) I would check for the error first and then the expected path as it mirrors the usual pattern in go to check for the error first